ur_controllers/src/gpio_controller.cpp

```diff
@@ -0,0 +1,326 @@
+  tool_data_msg_.analog_input_range3 =
+      static_cast<uint8_t>(state_interfaces_[StateInterfaces::TOOL_ANALOG_IO_TYPES + 1].get_value());
+  tool_data_msg_.analog_input2 = static_cast<float>(state_interfaces_[StateInterfaces::TOOL_ANALOG_INPUTS].get_value());
+  tool_data_msg_.analog_input3 =
+      static_cast<float>(state_interfaces_[StateInterfaces::TOOL_ANALOG_INPUTS + 1].get_value());
+  tool_data_msg_.tool_output_voltage =
+      static_cast<uint8_t>(state_interfaces_[StateInterfaces::TOOL_OUTPUT_VOLTAGE].get_value());
+  tool_data_msg_.tool_current = static_cast<float>(state_interfaces_[StateInterfaces::TOOL_OUTPUT_CURRENT].get_value());
+  tool_data_msg_.tool_temperature =
+      static_cast<float>(state_interfaces_[StateInterfaces::TOOL_TEMPERATURE].get_value());
+  tool_data_pub_->publish(tool_data_msg_);
+}
+
+void GPIOController::publishRobotMode()
+{
+  auto robot_mode = static_cast<int8_t>(state_interfaces_[StateInterfaces::ROBOT_MODE].get_value());
+
+  if (robot_mode_msg_.mode != robot_mode)
+  {
+    robot_mode_msg_.mode = robot_mode;
+    robot_mode_pub_->publish(robot_mode_msg_);
+  }
+}
+
+void GPIOController::publishSafetyMode()
+{
+  auto safety_mode = static_cast<uint8_t>(state_interfaces_[StateInterfaces::SAFETY_MODE].get_value());
+
+  if (safety_mode_msg_.mode != safety_mode)
+  {
+    safety_mode_msg_.mode = safety_mode;
+    safety_mode_pub_->publish(safety_mode_msg_);
+  }
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+ur_controllers::GPIOController::on_activate(const rclcpp_lifecycle::State& /*previous_state*/)
+{
+  return LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+ur_controllers::GPIOController::on_deactivate(const rclcpp_lifecycle::State& /*previous_state*/)
+{
+  return LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+bool GPIOController::setIO(ur_msgs::srv::SetIO::Request::SharedPtr req, ur_msgs::srv::SetIO::Response::SharedPtr resp)
+{
+  if (req->fun == req->FUN_SET_DIGITAL_OUT && req->pin >= 0 && req->pin <= 17)
+  {
+    // io async success
+    command_interfaces_[CommandInterfaces::IO_ASYNC_SUCCESS].set_value(2.0);
+    command_interfaces_[req->pin].set_value(static_cast<double>(req->state));
+
+    RCLCPP_INFO(node_->get_logger(), "Setting digital output '%d' to state: '%1.0f'.", req->pin, req->state);
+
+    while (command_interfaces_[CommandInterfaces::IO_ASYNC_SUCCESS].get_value() != 2.0)
+    {
+      // setting the value is not yet finished
+      std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    }
+
+    resp->success = static_cast<bool>(command_interfaces_[20].get_value());
+    return resp->success;
+  }
+  else if (req->fun == req->FUN_SET_ANALOG_OUT && req->pin >= 0 && req->pin <= 2)
+  {
+    // io async success
+    command_interfaces_[CommandInterfaces::IO_ASYNC_SUCCESS].set_value(2.0);
```

Comment:
Why is this value of 2.0 significant? It prob should be an enum

Reply:
It represents success. Enum would be better for sure...

Reply:
@livanov93 can you please take a look? It seems like 2.0 is a typo.

In gpio_controller.hpp, `20` is IO_ASYNC_SUCCESS. Not `2.0`